client/src/main/java/com/microsoft/durabletask/TaskOrchestrationExecutor.java

```
@@ -988,45 +993,77 @@ private RetriableTask(
                     TaskFactory<V> taskFactory,
                     @Nullable RetryPolicy retryPolicy,
                     @Nullable RetryHandler retryHandler) {
-                super(new CompletableFuture<>());
                 this.context = context;
                 this.taskFactory = taskFactory;
                 this.policy = retryPolicy;
                 this.handler = retryHandler;
                 this.firstAttempt = context.getCurrentInstant();
                 this.totalRetryTime = Duration.ZERO;
+                this.constructParentChildTask(taskFactory);
+            }
+
+            // Every RetriableTask will hava a CompletableTask as child.
+            private void constructParentChildTask(TaskFactory<V> taskFactory) {
+                Task<V> childTask = taskFactory.create();
+                this.setChildTask(childTask);
+                childTask.setParentTask(this);
             }
 
             @Override
-            public V await() {
-                Instant startTime = this.context.getCurrentInstant();
-                while (true) {
-                    Task<V> currentTask = this.taskFactory.create();
+            void handleChildSuccess(V result) {
+                this.complete(result);
+            }
 
-                    this.attemptNumber++;
+            @Override
+            void handleChildException(Throwable ex) {
+                if (ex instanceof TaskFailedException) {
+                    tryRetry((TaskFailedException) ex);
+                }
```

Should we throw an exception or log a warning if `ex` is not a `TaskFailedException`?

---

Good Catch, I review the code path again, I think for exception arrive here can only be `TaskFailedException`, so I plan to remove the if condition.